src/components/SignUpForm.js

```diff
@@ -0,0 +1,167 @@
+import React, { Component } from 'react'
+import zxcvbn from 'zxcvbn'
+import { validate } from '../helpers/validations'
+import { Button, Form, Label, Progress } from 'semantic-ui-react'
+import _ from 'lodash'
+
+const PasswordStrengthBar = ({ score }) =>
+  <Progress
+    attached="bottom"
+    title={`Password strength: ${[
+      'Very weak',
+      'Weak',
+      'Okay',
+      'Strong',
+      'Very strong'
+    ][score]}`}
+    style={{ marginTop: -3 }}
+    percent={score * 25}
+    color={['red', 'orange', 'yellow', 'olive', 'green'][score]}
+  />
+
+export class SignUpForm extends Component {
+  constructor(props) {
+    super(props)
+    this.fields = {}
+    this.state = {
+      passwordScore: 0,
+      errors: {}
+    }
+    this.onSubmit = this.onSubmit.bind(this)
+  }
+  componentDidMount() {
+    this._isMounted = true
+  }
+  componentWillUnmount() {
+    this._isMounted = false
+  }
+  fieldValues() {
+    return _.mapValues(this.fields, (f) => f && f.value)
+  }
+  onSubmit(event) {
+    event.preventDefault()
+    let fieldValues = this.fieldValues(),
+        validationErrors = validate(this.props.validations, fieldValues)
+    return validationErrors.then((errors) => {
+      this.setState({
+        errors: errors
+      })
+
+      if (Object.keys(errors).length === 0) {
+        return this.props
+          .onSubmit(fieldValues)
+          .then(this.props.onSuccess, this.props.onError)
+      } else {
+        return this.props.onError()
+      }
+    })
+  }
+  validateField(fieldName) {
+    return () =>
+      validate(
+        this.props.validations,
+        this.fieldValues(),
+        fieldName
+      ).then((errors) => {
+        if (this._isMounted) {
+          this.setState((state) => ({
+            ...state,
```

Comment:
I just tested if this line is redundant: It is.
```javascript
      ).then((errors) => {
        if (this._isMounted) {
          this.setState((state) => ({
            errors: { ...state.errors, ...errors }
          }))
        }
        console.log(this.state.passwordScore)
        console.log(this.state.errors)
      })
```
The passwordScore stays the same when I change the other fields (while introducing or removing validation errors).

So please remove this line :)